### PR TITLE
Refactoring NTO MachineConfig InPlace and Replace E2E Tests

### PR DESCRIPTION
**Pull request by jparrill**
Refactor of NTO MachineConfig for InPlace and Replace upgrade modes

- [x] Added refactored NTO MachineConfig test to the E2E suite
- [x] Modified current already refactored MachineConfig tests to avoid conflicting with other Nodes
- [x] Added `imagev1` to the API scheme (It was generating issues on data recollection after a test failure)

Signed-off-by: Juan Manuel Parrilla Madrid <jparrill@redhat.com>

**Which issue(s) this PR fixes**:

- Fixes #[HOSTEDCP-687](https://issues.redhat.com/browse/HOSTEDCP-687)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -258,7 +258,7 @@ func (o *options) DefaultClusterOptions(t *testing.T) core.CreateOptions {
 		ControlPlaneOperatorImage:        o.configurableClusterOptions.ControlPlaneOperatorImage,
 		ExternalDNSDomain:                o.configurableClusterOptions.ExternalDNSDomain,
 		AWSPlatform: core.AWSPlatformOptions{
-			InstanceType:       "m5.large",
+			InstanceType:       "m5.xlarge",
```
Comment (enxebre) [inline test/e2e/e2e_test.go:261]: why is this needed now and it was not before?

File: test/e2e/nodepool_machineconfig_test.go [hunk 2]
```diff
@@ -179,7 +179,7 @@ func testNodepoolMachineconfigGetsRolledout(parentCtx context.Context, mgmtClien
 				return false, nil
 			}
 
-			if len(pods.Items) != len(nodes) {
+			if len(pods.Items) < len(nodes) {
```
Comment (enxebre) [inline test/e2e/nodepool_machineconfig_test.go:182]: why relaxing this?

File: test/e2e/nodepool_nto_machineconfig_test.go [hunk 3]
```diff
@@ -46,222 +50,281 @@ spec:
...
 				return false, nil
 			}
-		}
 
-		return true, nil
-	})
-	if err != nil {
-		t.Fatalf("failed waiting for all pods in the NTO MachineConfig update verification DS to be ready: %v", err)
-	}
+			if len(pods.Items) < len(nodes) {
```
Comment (enxebre) [inline test/e2e/nodepool_nto_machineconfig_test.go:308]: why relaxing this?

File: test/e2e/util/util.go [hunk 4]
```diff
@@ -370,8 +370,11 @@ func EnsureNoCrashingPods(t *testing.T, ctx context.Context, client crclient.Cli
 			}
 
 			for _, containerStatus := range pod.Status.ContainerStatuses {
-				if containerStatus.RestartCount > 0 {
-					t.Errorf("Container %s in pod %s has a restartCount > 0 (%d)", containerStatus.Name, pod.Name, containerStatus.RestartCount)
+				// Setting this to 5 just to cover external controllers and operators that
+				// need some tries to run properly until other components are accesible
+				// E.G csi-snapshot-controller
```
Comment (enxebre) [inline test/e2e/util/util.go:375]: We need to fix crash loops.
If we wanted to temporary relax, let's put a PR for that aside from this one.

**Comment by enxebre** [comment]
thanks! this is looking great. Dropped some questions

File: test/e2e/nodepool_nto_machineconfig_test.go (hunk 3 above)
Comment (jparrill) [reply test/e2e/nodepool_nto_machineconfig_test.go:308]: because they share the same HostedCluster and they have the same labels. For now this is temporary measure until I finish the PR (Still not there). I've submitted it to share thoughts with Cesar and you ;). The plan that I've though is to add a Label to identify it as unique and set the older one.

File: test/e2e/nodepool_machineconfig_test.go (hunk 2 above)
Comment (jparrill) [reply test/e2e/nodepool_machineconfig_test.go:182]: Answered down below

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (jparrill) [reply test/e2e/e2e_test.go:261]: I'm still testing but the example HC we create now manages 1-N NodePools, so eventually this HC will need some boost. The plan is to check if the older one is enough to handle all 4 tests and if not we can continue the discussion

File: test/e2e/util/util.go (hunk 4 above)
Comment (jparrill) [reply test/e2e/util/util.go:375]: Agree, is a temporary measure, I will change before the last Review ;)

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (enxebre) [reply test/e2e/e2e_test.go:261]: > HC we create now manages 1-N NodePools,
I'm not sure how this relates to increase the size of the the NodePool instances

File: test/e2e/nodepool_nto_machineconfig_test.go (hunk 3 above)
Comment (enxebre) [reply test/e2e/nodepool_nto_machineconfig_test.go:308]: We know which NodePool the test is operating against. We know which Nodes (they have a NodePool label) belong to a NodePool. So I'm not seeing the need to relax comparing to current approach.

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (jparrill) [reply test/e2e/e2e_test.go:261]: Woah you're right, brain fart... I will update the PR soon (after the Tests execution in local)

**Comment by enxebre** [comment]
Please squash commits and let's use [imperative mood for message](https://github.com/openshift/hypershift/blob/main/docs/content/contribute/index.md)

File: test/e2e/nodepool_machineconfig_test.go [hunk 5]
```diff
@@ -163,13 +163,35 @@ func testNodepoolMachineconfigGetsRolledout(parentCtx context.Context, mgmtClien
 			t.Fatalf("failed to update nodepool %s after adding machineconfig: %v", nodePool.Name, err)
 		}
 
+		// DS Customization
```
Comment (enxebre) [inline test/e2e/nodepool_machineconfig_test.go:166]: I thought this PR is for n NTO, how is this change related?

Comment (jparrill) [reply test/e2e/nodepool_machineconfig_test.go:166]: Just to follow the same pattern in every similar test. I can come back to the older one, but this looks better if you will add more tests like this in the future.

File: test/e2e/nodepool_nto_machineconfig_test.go [hunk 6]
```diff
@@ -46,222 +50,335 @@ spec:
...
+				t.Fatalf("failed to create nodePool %s with Autorepair function: %v", nodePool.Name, err)
+			}
+			err = nodePoolRecreate(t, ctx, nodePool, mgmtClient)
+			g.Expect(err).NotTo(HaveOccurred(), "failed to Create the NodePool")
+		}
+		defer nodePoolScaleDownToZero(ctx, mgmtClient, *nodePool, t)
 
-	// Wait for Nodes to be Ready
-	e2eutil.WaitForNReadyNodes(t, testContext, guestClient, clusterOpts.NodePoolReplicas, hostedCluster.Spec.Platform.Type)
+		numNodes := int32(1)
```
Comment (enxebre) [inline test/e2e/nodepool_nto_machineconfig_test.go:124]: why is this 1 now? what was is before? I'd expect it to be at least 2

File: test/e2e/nodepool_nto_machineconfig_test.go [hunk 7]
```diff
@@ -46,222 +50,335 @@ spec:
...
-		for _, pod := range pods.Items {
-			if !isPodReady(&pod) {
-				return false, nil
+		ds.Name = dsName
+		ds.ObjectMeta.Labels = make(map[string]string)
+		ds.ObjectMeta.Labels["hypershift.openshift.io/nodePool"] = nodePool.Name
+
+		ds.Spec.Selector.MatchLabels["name"] = dsName
+		ds.Spec.Template.ObjectMeta.Labels["name"] = dsName
+		ds.Spec.Template.ObjectMeta.Labels["hypershift.openshift.io/nodePool"] = nodePool.Name
```
Comment (enxebre) [inline test/e2e/nodepool_nto_machineconfig_test.go:170]: why don't use hypershift.openshift.io/nodePool in MatchLabels as well?

**Comment by enxebre** [comment]
thanks! dropped some questions which also apply for testNTOMachineConfigAppliedInPlace

File: test/e2e/nodepool_nto_machineconfig_test.go (hunk 6 above)
Comment (jparrill) [reply test/e2e/nodepool_nto_machineconfig_test.go:124]: only in the Replace case? or also for InPlace?

File: test/e2e/nodepool_machineconfig_test.go (hunk 5 above)
Comment (jparrill) [reply test/e2e/nodepool_machineconfig_test.go:166]: Also, the way it is now, conflicts with the current approach in NTO, this is why I'm "fixing" the MC tests

File: test/e2e/nodepool_nto_machineconfig_test.go (hunk 7 above)
Comment (jparrill) [reply test/e2e/nodepool_nto_machineconfig_test.go:170]: Makes totally sense

File: test/e2e/nodepool_nto_machineconfig_test.go (hunk 6 above)
Comment (jparrill) [reply test/e2e/nodepool_nto_machineconfig_test.go:124]: Modified in both cases (InPlace and Replace)

File: test/e2e/nodepool_nto_machineconfig_test.go [hunk 8]
```diff
@@ -46,222 +50,335 @@ spec:
...
-	t.Parallel()
-	g := NewWithT(t)
+			if podCount != len(nodes) {
+				return false, nil
+			}
 
-	ctx, cancel := context.WithCancel(testContext)
-	defer cancel()
+			for _, pod := range pods.Items {
+				if pod.Labels["hypershift.openshift.io/nodePool"] == nodePool.Name && !isPodReady(&pod) {
```
[the remark this answers is not in the document]
Comment (jparrill) [reply test/e2e/nodepool_nto_machineconfig_test.go]: I've modified also the MachineConfig one.

File: test/e2e/nodepool_nto_machineconfig_test.go [hunk 9]
```diff
@@ -46,222 +50,325 @@ spec:
...
+				t.Fatalf("failed to create nodePool %s with Autorepair function: %v", nodePool.Name, err)
+			}
+			err = nodePoolRecreate(t, ctx, nodePool, mgmtClient)
+			g.Expect(err).NotTo(HaveOccurred(), "failed to Create the NodePool")
+		}
+		defer nodePoolScaleDownToZero(ctx, mgmtClient, *nodePool, t)
 
-	// Wait for Nodes to be Ready
-	e2eutil.WaitForNReadyNodes(t, testContext, guestClient, clusterOpts.NodePoolReplicas, hostedCluster.Spec.Platform.Type)
+		numNodes := int32(2)
```
Comment (enxebre) [inline test/e2e/nodepool_nto_machineconfig_test.go:124]: not sure why we use this variable here instead of reusing `twoReplicas`?

**Comment by enxebre** [comment]
can you please squash commits and clean up message again?
Thanks!
/approve

File: api/scheme.go [hunk 10]
```diff
@@ -5,6 +5,7 @@ import (
 
 	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
 	configv1 "github.com/openshift/api/config/v1"
+	imagev1 "github.com/openshift/api/image/v1"
```
Comment (enxebre) [inline api/scheme.go:8]: why this change?

Comment (jparrill) [reply api/scheme.go:8]: There was an issue on data recolection after a failure and discussing with @csrwng we agreed in add it to the scheme.

Comment (csrwng) [reply api/scheme.go:8]: @jparrill that was added by this PR (https://github.com/openshift/hypershift/pull/2054), so you don't need to carry it here. Actually if you rebase your branch on main, it should go away
